File: ruby/sum-of-multiples/sum_of_multiples.rb
class SumOfMultiples
  def initialize(*nums)
    @nums = nums
  end

  def to(i)
    @nums.each_with_object(Array.new) do |n, multiples|
      n_multiple = n
      while n_multiple < i
        multiples << n_multiple
        n_multiple += n
      end
    end.uniq.reduce(0, :+)
  end
end

File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    Enum.reduce(
      (for n <- 1..limit-1, Enum.any?(factors, &(rem(n, &1) == 0)), do: n),
      0,
      fn(x, acc) -> x + acc end
    )
  end
end
